lib/spack/spack/build_environment.py

```diff
@@ -993,8 +1011,26 @@ def add_modifications_for_dep(dep):
             dpkg = dep.package
             if set_package_py_globals:
                 set_module_variables_for_package(dpkg)
+
             # Allow dependencies to modify the module
-            dpkg.setup_dependent_package(spec.package.module, spec)
+            # Get list of modules that may need updating
+            modules = []
+            for cls in inspect.getmro(type(spec.package)):
+                module = cls.module
+                if module == spack.package_base:
+                    break
+                modules.append(module)
+
+            # Execute changes as if on a single module
+            # copy dict to ensure prior changes are available
+            module_accumulator = ModuleChangeAccumulator()
+            module_accumulator.__dict__ = spec.package.module.__dict__
+            dpkg.setup_dependent_package(module_accumulator, spec)
+
+            for module in modules:
+                for key, value in module_accumulator.changes.items():
+                    module.__dict__[key] = value  # setattr(module, _) python 3.7+
```

Comment:
Pretty sure this works:

```suggestion
            changes = spack.util.pattern.Bunch(spec.package.module.__dict__)
            dpkg.setup_dependent_package(changes, spec)

            for module in modules:
                module.__dict__.update(spec.package.module.__dict__)
```